File: packages/sandboxd/src/pi_proxy/json_rpc.rs
//! JSON-RPC request handling for the Pi websocket proxy.
//!
//! This module validates websocket messages, maps supported RPC methods onto
//! session/process operations, and keeps method-specific response formatting out
//! of the socket loop.

use std::sync::Arc;

use serde::{Deserialize, Serialize};
use serde_json::{Value, json};

use crate::pi_proxy::{PiProxyError, PiProxyState, session};

#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
struct JsonRpcRequest {
    id: Value,
    method: String,
    params: Option<Value>,
}

#[derive(Debug, Serialize)]
struct JsonRpcError {
    code: i64,
    message: String,
}

#[derive(Debug, Serialize)]
struct JsonRpcErrorResponse {
    jsonrpc: &'static str,
    id: Value,
    error: JsonRpcError,
}

#[derive(Debug, Serialize)]
struct JsonRpcSuccessResponse {
    jsonrpc: &'static str,
    id: Value,
    result: Value,
}

pub(super) fn handle_json_rpc_request(state: &Arc<PiProxyState>, payload: &str) -> Vec<String> {
    let request = match serde_json::from_str::<JsonRpcRequest>(payload) {
        Ok(request) => request,
        Err(error) => {
            return vec![render_json_rpc_error(
                Value::Null,
                -32_700,
                format!("Invalid JSON-RPC request: {error}"),
            )];
        }
    };
    let mut captured_events = Vec::new();
    let result = match handle_pi_method(state, &request, &mut captured_events) {
        Ok(result) => JsonRpcSuccessResponse {
            jsonrpc: "2.0",
            id: request.id,
            result,
        },
        Err(error) => {
            return vec![render_json_rpc_error(
                request.id,
                -32_000,
                error.to_string(),
            )];
        }
    };
    let mut responses: Vec<String> = captured_events
        .into_iter()
        .map(render_pi_event_json_rpc_notification)
        .collect();
    match serde_json::to_string(&result) {
        Ok(response) => responses.push(response),
        Err(error) => responses.push(render_json_rpc_error(
            Value::Null,
            -32_000,
            error.to_string(),
        )),
    }
    responses
}

pub(super) fn render_pi_event_json_rpc_notification(event: Value) -> String {
    serde_json::to_string(&json!({
        "jsonrpc": "2.0",
        "method": "pi/event",
        "params": event,
    }))
    .unwrap_or_else(|_| {
        "{\"jsonrpc\":\"2.0\",\"method\":\"pi/event\",\"params\":{\"type\":\"serialization_error\"}}"
            .to_string()
    })
}

fn render_json_rpc_error(id: Value, code: i64, message: String) -> String {
    let response = JsonRpcErrorResponse {
        jsonrpc: "2.0",
        id,
        error: JsonRpcError { code, message },
    };
    serde_json::to_string(&response).unwrap_or_else(|_| {
        "{\"jsonrpc\":\"2.0\",\"id\":null,\"error\":{\"code\":-32000,\"message\":\"failed to serialize error\"}}"
            .to_string()
    })
}

fn read_param_string(params: &Option<Value>, key: &str) -> Option<String> {
    params
        .as_ref()
        .and_then(|value| value.get(key))
        .and_then(Value::as_str)
        .map(ToString::to_string)
}

fn read_param_usize(params: &Option<Value>, key: &str) -> Option<usize> {
    params
        .as_ref()
        .and_then(|value| value.get(key))
        .and_then(Value::as_u64)
        .and_then(|value| usize::try_from(value).ok())
}

fn require_param_usize(params: &Option<Value>, key: &str) -> Result<usize, PiProxyError> {
    read_param_usize(params, key)
        .ok_or_else(|| PiProxyError::InvalidRequest(format!("missing required parameter '{key}'")))
}

fn require_param_string(params: &Option<Value>, key: &str) -> Result<String, PiProxyError> {
    read_param_string(params, key)
        .ok_or_else(|| PiProxyError::InvalidRequest(format!("missing required parameter '{key}'")))
}

fn handle_pi_method(
    state: &Arc<PiProxyState>,
    request: &JsonRpcRequest,
    captured_events: &mut Vec<Value>,
) -> Result<Value, PiProxyError> {
    match request.method.as_str() {
        "pi/createConversation" => {
            let cwd = read_param_string(&request.params, "cwd");
            state.ensure_child(cwd.as_deref())?;
            state.send_pi_command_with_captured_events(
                json!({ "type": "new_session" }),
                captured_events,
            )?;
            let state_value = state.send_pi_command_with_captured_events(
                json!({ "type": "get_state" }),
                captured_events,
            )?;
            let session_file = PiProxyState::read_session_file(&state_value)?;
            let provider_conversation_id = state_value["sessionId"].as_str().ok_or_else(|| {
                PiProxyError::InvalidRequest("Pi did not report sessionId".to_string())
            })?;
            Ok(json!({
                "providerConversationId": provider_conversation_id,
                "sessionFile": session_file,
            }))
        }
        "pi/findRecentConversation" => {
            let cwd = read_param_string(&request.params, "cwd");
            let conversation =
                session::find_recent_conversation(&state.config.env, cwd.as_deref())?;
            Ok(match conversation {
                Some(conversation) => json!({
                    "providerConversationId": conversation.id,
                }),
                None => json!({
                    "providerConversationId": Value::Null,
                }),
            })
        }
        "pi/listConversations" => {
            let cwd = read_param_string(&request.params, "cwd");
            let limit = require_param_usize(&request.params, "limit")?;
            session::list_conversations(&state.config.env, cwd.as_deref(), limit)
        }
        "pi/resolveConversation" => {
            let provider_conversation_id =
                require_param_string(&request.params, "providerConversationId")?;
            let conversation =
                session::find_conversation_by_id(&state.config.env, &provider_conversation_id)?;
            Ok(json!({ "sessionFile": conversation.path.to_string_lossy().to_string() }))
        }
        "pi/getState" => {
            let session_file = read_param_string(&request.params, "sessionFile");
            state.ensure_child(None)?;
            if let Some(session_file) = session_file {
                state.switch_session(&session_file, captured_events)?;
            }
            state.send_pi_command_with_captured_events(
                json!({ "type": "get_state" }),
                captured_events,
            )
        }
        "pi/getAvailableModels" => {
            let session_file = require_param_string(&request.params, "sessionFile")?;
            state.ensure_child(None)?;
            state.switch_session(&session_file, captured_events)?;
            state.send_pi_command_with_captured_events(
                json!({ "type": "get_available_models" }),
                captured_events,
            )
        }
        "pi/readMetadata" => {
            let session_file = require_param_string(&request.params, "sessionFile")?;
            state.ensure_child(None)?;
            state.switch_session(&session_file, captured_events)?;
            let state_value = state.send_pi_command_with_captured_events(
                json!({ "type": "get_state" }),
                captured_events,
            )?;
            Ok(json!({
                "name": state_value.get("sessionName").cloned().unwrap_or(Value::Null),
                "preview": Value::Null
            }))
        }
        "pi/getMessages" => {
            let session_file = require_param_string(&request.params, "sessionFile")?;
            state.ensure_child(None)?;
            state.switch_session(&session_file, captured_events)?;
            let messages_value = state.send_pi_command_with_captured_events(
                json!({ "type": "get_messages" }),
                captured_events,
            )?;
            Ok(messages_value)
        }
        "pi/resumeConversation" => {
            let provider_conversation_id =
                require_param_string(&request.params, "providerConversationId")?;
            let conversation =
                session::find_conversation_by_id(&state.config.env, &provider_conversation_id)?;
            let session_file = conversation.path.to_string_lossy().to_string();
            state.ensure_child(None)?;
            state.switch_session(&session_file, captured_events)?;
            PiProxyState::mark_active_and_start_activity_monitor(state);
            Ok(json!({ "sessionFile": session_file }))
        }
        "pi/setModel" => {
            let session_file = require_param_string(&request.params, "sessionFile")?;
            let provider = require_param_string(&request.params, "provider")?;
            let model_id = require_param_string(&request.params, "modelId")?;
            state.ensure_child(None)?;
            state.switch_session(&session_file, captured_events)?;
            state.send_pi_command_with_captured_events(
                json!({ "type": "set_model", "provider": provider, "modelId": model_id }),
                captured_events,
            )
        }
        "pi/setThinkingLevel" => {
            let session_file = require_param_string(&request.params, "sessionFile")?;
            let level = require_param_string(&request.params, "level")?;
            state.ensure_child(None)?;
            state.switch_session(&session_file, captured_events)?;
            state.send_pi_command_with_captured_events(
                json!({ "type": "set_thinking_level", "level": level }),
                captured_events,
            )
        }
        "pi/setSessionName" => {
            let session_file = require_param_string(&request.params, "sessionFile")?;
            let name = require_param_string(&request.params, "name")?;
            state.ensure_child(None)?;
            state.switch_session(&session_file, captured_events)?;
            state.send_pi_command_with_captured_events(
                json!({ "type": "set_session_name", "name": name }),
                captured_events,
            )
        }
        "pi/prompt" => {
            let session_file = require_param_string(&request.params, "sessionFile")?;
            let message = require_param_string(&request.params, "message")?;
            state.ensure_child(None)?;
            state.switch_session(&session_file, captured_events)?;
            PiProxyState::mark_active_and_start_activity_monitor(state);
            state.send_pi_command_with_captured_events(
                json!({ "type": "prompt", "message": message }),
                captured_events,
            )
        }
        "pi/steer" => {
            let session_file = require_param_string(&request.params, "sessionFile")?;
            let message = require_param_string(&request.params, "message")?;
            state.ensure_child(None)?;
            state.switch_session(&session_file, captured_events)?;
            PiProxyState::mark_active_and_start_activity_monitor(state);
            state.send_pi_command_with_captured_events(
                json!({ "type": "steer", "message": message }),
                captured_events,
            )
        }
        "pi/followUp" => {
            let session_file = require_param_string(&request.params, "sessionFile")?;
            let message = require_param_string(&request.params, "message")?;
            state.ensure_child(None)?;
            state.switch_session(&session_file, captured_events)?;
            PiProxyState::mark_active_and_start_activity_monitor(state);
            state.send_pi_command_with_captured_events(
                json!({ "type": "follow_up", "message": message }),
                captured_events,
            )
        }
        "pi/abort" => {
            let session_file = require_param_string(&request.params, "sessionFile")?;
            state.ensure_child(None)?;
            state.switch_session(&session_file, captured_events)?;
            state.send_pi_command_with_captured_events(json!({ "type": "abort" }), captured_events)
        }
        other => Err(PiProxyError::InvalidRequest(format!(
            "unsupported Pi proxy method '{other}'"
        ))),
    }
}

File: packages/sandboxd/piproxy/jsonrpc.go
package piproxy

import (
	"bytes"
	"encoding/json"
	"fmt"
)

type jsonRPCRequest struct {
	ID          any             `json:"id"`
	Method      string          `json:"method"`
	Params      map[string]any  `json:"params"`
	Idempotency json.RawMessage `json:"idempotency"`
}

type jsonRPCError struct {
	Code    int64  `json:"code"`
	Message string `json:"message"`
}

func HandleJSONRPCRequest(state *State, payload []byte) [][]byte {
	request, err := decodeJSONRPCRequest(payload)
	if err != nil {
		return [][]byte{renderJSONRPCError(request.ID, -32700, fmt.Sprintf("Invalid JSON-RPC request: %v", err))}
	}
	idempotencyAction := prepareIdempotency(request, state.idempotencyStore)
	switch idempotencyAction.kind {
	case idempotencyActionDisabled:
	case idempotencyActionForward:
	case idempotencyActionReplay:
		payload := idempotencyAction.replay.Payload
		payload["id"] = request.ID
		return [][]byte{mustMarshalJSON(payload)}
	case idempotencyActionReject:
		return [][]byte{renderJSONRPCError(request.ID, -32001, idempotencyAction.message)}
	default:
		return [][]byte{renderJSONRPCError(request.ID, -32001, fmt.Sprintf("unsupported Pi idempotency action %q", idempotencyAction.kind))}
	}
	var capturedEvents []map[string]any
	result, err := handlePiMethod(state, request, &capturedEvents)
	if err != nil {
		response := jsonRPCErrorPayload(request.ID, -32000, err.Error())
		if idempotencyAction.started != nil {
			if completeErr := completeIdempotency(state.idempotencyStore, *idempotencyAction.started, StoredResponse{Payload: response}); completeErr != nil {
				return [][]byte{renderJSONRPCError(nil, -32000, completeErr.Error())}
			}
		}
		return [][]byte{mustMarshalJSON(response)}
	}
	responses := make([][]byte, 0, len(capturedEvents)+1)
	for _, event := range capturedEvents {
		responses = append(responses, RenderPiEventJSONRPCNotification(event))
	}
	response := jsonRPCSuccessPayload(request.ID, result)
	if idempotencyAction.started != nil {
		if err := completeIdempotency(state.idempotencyStore, *idempotencyAction.started, StoredResponse{Payload: response}); err != nil {
			responses = append(responses, renderJSONRPCError(nil, -32000, err.Error()))
			return responses
		}
	}
	responses = append(responses, mustMarshalJSON(response))
	return responses
}

func decodeJSONRPCRequest(payload []byte) (jsonRPCRequest, error) {
	var raw map[string]json.RawMessage
	if err := json.Unmarshal(payload, &raw); err != nil {
		return jsonRPCRequest{}, err
	}
	request := jsonRPCRequest{Params: nil}
	if rawID, ok := raw["id"]; ok {
		if err := json.Unmarshal(rawID, &request.ID); err != nil {
			return request, err
		}
	} else {
		return request, fmt.Errorf("missing field `id`")
	}
	rawMethod, ok := raw["method"]
	if !ok {
		return request, fmt.Errorf("missing field `method`")
	}
	if err := json.Unmarshal(rawMethod, &request.Method); err != nil {
		return request, err
	}
	if rawParams, ok := raw["params"]; ok && !bytes.Equal(bytes.TrimSpace(rawParams), []byte("null")) {
		var paramsValue any
		if err := json.Unmarshal(rawParams, &paramsValue); err != nil {
			return request, err
		}
		if params, ok := paramsValue.(map[string]any); ok {
			request.Params = params
		}
	}
	if rawIdempotency, ok := raw["idempotency"]; ok {
		request.Idempotency = append(json.RawMessage(nil), rawIdempotency...)
	}
	return request, nil
}

func decodeStrictPiIdempotency(payload []byte, target *Idempotency) error {
	decoder := json.NewDecoder(bytes.NewReader(payload))
	decoder.DisallowUnknownFields()
	return decoder.Decode(target)
}

func RenderPiEventJSONRPCNotification(event map[string]any) []byte {
	return mustMarshalJSON(map[string]any{
		"jsonrpc": "2.0",
		"method":  "pi/event",
		"params":  event,
	})
}

func renderJSONRPCSuccess(id any, result any) []byte {
	return mustMarshalJSON(jsonRPCSuccessPayload(id, result))
}

func jsonRPCSuccessPayload(id any, result any) map[string]any {
	return map[string]any{
		"jsonrpc": "2.0",
		"id":      id,
		"result":  result,
	}
}

func renderJSONRPCError(id any, code int64, message string) []byte {
	return mustMarshalJSON(jsonRPCErrorPayload(id, code, message))
}

func jsonRPCErrorPayload(id any, code int64, message string) map[string]any {
	return map[string]any{
		"jsonrpc": "2.0",
		"id":      id,
		"error": jsonRPCError{
			Code:    code,
			Message: message,
		},
	}
}

func handlePiMethod(state *State, request jsonRPCRequest, capturedEvents *[]map[string]any) (any, error) {
	switch request.Method {
	case "pi/createConversation":
		cwd := optionalStringParam(request.Params, "cwd")
		if err := state.EnsureChild(cwd); err != nil {
			return nil, err
		}
		if _, err := state.SendCommandWithCapturedEvents(map[string]any{"type": "new_session"}, capturedEvents); err != nil {
			return nil, err
		}
		stateValue, err := state.SendCommandWithCapturedEvents(map[string]any{"type": "get_state"}, capturedEvents)
		if err != nil {
			return nil, err
		}
		sessionFile, err := readSessionFileFromState(stateValue)
		if err != nil {
			return nil, err
		}
		providerConversationID, ok := stateValue["sessionId"].(string)
		if !ok || providerConversationID == "" {
			return nil, fmt.Errorf("Pi did not report sessionId")
		}
		return map[string]any{
			"providerConversationId": providerConversationID,
			"sessionFile":            sessionFile,
		}, nil
	case "pi/findRecentConversation":
		conversation, err := FindRecentConversation(state.config.Env, optionalStringParam(request.Params, "cwd"))
		if err != nil {
			return nil, err
		}
		if conversation == nil {
			return map[string]any{"providerConversationId": nil}, nil
		}
		return map[string]any{"providerConversationId": conversation.ID}, nil
	case "pi/listConversations":
		limit, err := requiredIntParam(request.Params, "limit")
		if err != nil {
			return nil, err
		}
		return ListConversations(state.config.Env, optionalStringParam(request.Params, "cwd"), limit)
	case "pi/resolveConversation":
		providerConversationID, err := requiredStringParam(request.Params, "providerConversationId")
		if err != nil {
			return nil, err
		}
		conversation, err := FindConversationByID(state.config.Env, providerConversationID)
		if err != nil {
			return nil, err
		}
		return map[string]any{"sessionFile": conversation.Path}, nil
	case "pi/getState":
		if err := state.EnsureChild(nil); err != nil {
			return nil, err
		}
		if sessionFile := optionalStringParam(request.Params, "sessionFile"); sessionFile != nil {
			if err := state.SwitchSession(*sessionFile, capturedEvents); err != nil {
				return nil, err
			}
		}
		return state.SendCommandWithCapturedEvents(map[string]any{"type": "get_state"}, capturedEvents)
	case "pi/getAvailableModels":
		sessionFile, err := requiredStringParam(request.Params, "sessionFile")
		if err != nil {
			return nil, err
		}
		if err := state.EnsureChild(nil); err != nil {
			return nil, err
		}
		if err := state.SwitchSession(sessionFile, capturedEvents); err != nil {
			return nil, err
		}
		return state.SendCommandWithCapturedEvents(map[string]any{"type": "get_available_models"}, capturedEvents)
	case "pi/readMetadata":
		sessionFile, err := requiredStringParam(request.Params, "sessionFile")
		if err != nil {
			return nil, err
		}
		if err := state.EnsureChild(nil); err != nil {
			return nil, err
		}
		if err := state.SwitchSession(sessionFile, capturedEvents); err != nil {
			return nil, err
		}
		stateValue, err := state.SendCommandWithCapturedEvents(map[string]any{"type": "get_state"}, capturedEvents)
		if err != nil {
			return nil, err
		}
		return map[string]any{"name": stateValue["sessionName"], "preview": nil}, nil
	case "pi/getMessages":
		sessionFile, err := requiredStringParam(request.Params, "sessionFile")
		if err != nil {
			return nil, err
		}
		if err := state.EnsureChild(nil); err != nil {
			return nil, err
		}
		if err := state.SwitchSession(sessionFile, capturedEvents); err != nil {
			return nil, err
		}
		return state.SendCommandWithCapturedEvents(map[string]any{"type": "get_messages"}, capturedEvents)
	case "pi/resumeConversation":
		providerConversationID, err := requiredStringParam(request.Params, "providerConversationId")
		if err != nil {
			return nil, err
		}
		conversation, err := FindConversationByID(state.config.Env, providerConversationID)
		if err != nil {
			return nil, err
		}
		if err := state.EnsureChild(nil); err != nil {
			return nil, err
		}
		if err := state.SwitchSession(conversation.Path, capturedEvents); err != nil {
			return nil, err
		}
		state.MarkActiveAndStartActivityMonitor()
		return map[string]any{"sessionFile": conversation.Path}, nil
	case "pi/setModel":
		return stateCommandWithSession(state, request.Params, capturedEvents, "set_model", map[string]string{"provider": "provider", "modelId": "modelId"})
	case "pi/setThinkingLevel":
		return stateCommandWithSession(state, request.Params, capturedEvents, "set_thinking_level", map[string]string{"level": "level"})
	case "pi/setSessionName":
		return stateCommandWithSession(state, request.Params, capturedEvents, "set_session_name", map[string]string{"name": "name"})
	case "pi/prompt":
		return stateCommandWithSession(state, request.Params, capturedEvents, "prompt", map[string]string{"message": "message"})
	case "pi/steer":
		return stateCommandWithSession(state, request.Params, capturedEvents, "steer", map[string]string{"message": "message"})
	case "pi/followUp":
		return stateCommandWithSession(state, request.Params, capturedEvents, "follow_up", map[string]string{"message": "message"})
	case "pi/abort":
		return stateCommandWithSession(state, request.Params, capturedEvents, "abort", nil)
	default:
		return nil, fmt.Errorf("unsupported Pi method %q", request.Method)
	}
}

func stateCommandWithSession(
	state *State,
	params map[string]any,
	capturedEvents *[]map[string]any,
	commandType string,
	paramMappings map[string]string,
) (any, error) {
	sessionFile, err := requiredStringParam(params, "sessionFile")
	if err != nil {
		return nil, err
	}
	if err := state.EnsureChild(nil); err != nil {
		return nil, err
	}
	if err := state.SwitchSession(sessionFile, capturedEvents); err != nil {
		return nil, err
	}
	command := map[string]any{"type": commandType}
	for commandKey, paramKey := range paramMappings {
		value, err := requiredStringParam(params, paramKey)
		if err != nil {
			return nil, err
		}
		command[commandKey] = value
	}
	result, err := state.SendCommandWithCapturedEvents(command, capturedEvents)
	if err == nil && (commandType == "prompt" || commandType == "steer" || commandType == "follow_up") {
		state.MarkActiveAndStartActivityMonitor()
	}
	return result, err
}

func optionalStringParam(params map[string]any, key string) *string {
	if params == nil {
		return nil
	}
	value, ok := params[key].(string)
	if !ok {
		return nil
	}
	return &value
}

func requiredStringParam(params map[string]any, key string) (string, error) {
	value := optionalStringParam(params, key)
	if value == nil {
		return "", fmt.Errorf("missing required parameter %q", key)
	}
	return *value, nil
}

func requiredIntParam(params map[string]any, key string) (int, error) {
	if params == nil {
		return 0, fmt.Errorf("missing required parameter %q", key)
	}
	value, ok := params[key].(float64)
	if !ok {
		return 0, fmt.Errorf("missing required parameter %q", key)
	}
	return int(value), nil
}

func readSessionFileFromState(stateValue map[string]any) (string, error) {
	sessionFile, ok := stateValue["sessionFile"].(string)
	if !ok || sessionFile == "" {
		return "", fmt.Errorf("Pi did not report sessionFile")
	}
	return sessionFile, nil
}

func mustMarshalJSON(value any) []byte {
	serialized, err := json.Marshal(value)
	if err != nil {
		return []byte(`{"jsonrpc":"2.0","id":null,"error":{"code":-32000,"message":"failed to serialize response"}}`)
	}
	return serialized
}
